File: rust/crates/oxa-ir/src/checker.rs
//! Event-stream invariant checking (spec/01 §7): INV-5 grammar, INV-6 index
//! discipline, and the relational rule that a tool block's `input` equals
//! the exact concatenation of its `input_json_delta.partial_json` fragments
//! (INV-1: compared as opaque strings, never parsed).

use crate::event::{Delta, Event, EventStream};
use crate::request::Block;
use crate::response::StopReason;

/// A rule violation found by the event-stream validators.
#[derive(Clone, Debug, PartialEq)]
pub struct Violation {
    /// Index of the offending event (or the stream length for an EOF
    /// violation such as a missing terminal event).
    pub event: usize,
    pub message: String,
}

fn violate(event: usize, message: impl Into<String>) -> Violation {
    Violation {
        event,
        message: message.into(),
    }
}

/// Validates a decoder-produced stream (strict): a tool block with a
/// non-empty input must carry explicit input fragments.
pub fn validate_event_stream(stream: &EventStream) -> Result<(), Violation> {
    validate_with(stream, false)
}

/// Validates an encoder-input stream (lenient): accepts the documented
/// encoder shorthand where a tool block carries its full input and no
/// fragments (the encoder synthesizes one full delta).
pub fn validate_event_stream_for_encoder(stream: &EventStream) -> Result<(), Violation> {
    validate_with(stream, true)
}

struct OpenTool {
    input: String,
    fragments: String,
    fragment_count: usize,
}

enum OpenBlock {
    Text,
    Tool(OpenTool),
}

#[derive(Clone, Copy, PartialEq)]
enum Phase {
    NeedStart,
    Blocks,
    NeedMessageDone,
    Done,
}

fn validate_with(stream: &EventStream, allow_synthesized: bool) -> Result<(), Violation> {
    let mut phase = Phase::NeedStart;
    let mut next_index: i64 = 0;
    // The currently open block: (index, kind). INV-5 allows at most one.
    let mut open: Option<(i64, OpenBlock)> = None;

    for (i, event) in stream.events.iter().enumerate() {
        match event {
            Event::MessageStart { .. } => {
                if phase != Phase::NeedStart {
                    return Err(violate(i, "duplicate message_start"));
                }
                phase = Phase::Blocks;
            }
            Event::ContentBlockStart { index, block } => {
                if phase != Phase::Blocks {
                    return Err(violate(
                        i,
                        "content_block_start outside the block region (no message_start, or a message_delta already seen)",
                    ));
                }
                if open.is_some() {
                    return Err(violate(i, "content_block_start with an open block"));
                }
                if *index != next_index {
                    return Err(violate(
                        i,
                        format!("content_block_start index {index}, want {next_index}"),
                    ));
                }
                next_index += 1;
                open = Some(match block {
                    Block::Text { .. } => (*index, OpenBlock::Text),
                    Block::ToolUse { input, .. } => (
                        *index,
                        OpenBlock::Tool(OpenTool {
                            input: input.clone(),
                            fragments: String::new(),
                            fragment_count: 0,
                        }),
                    ),
                    other => {
                        return Err(violate(
                            i,
                            format!(
                                "content_block_start carries {} block; streams carry text and tool_use only",
                                block_kind(other)
                            ),
                        ));
                    }
                });
            }
            Event::ContentBlockDelta { index, delta } => {
                let Some((open_index, kind)) = open.as_mut() else {
                    return Err(violate(i, "content_block_delta without an open block"));
                };
                if *index != *open_index {
                    return Err(violate(
                        i,
                        format!(
                            "content_block_delta index {index} does not match the open block {open_index}"
                        ),
                    ));
                }
                match (kind, delta) {
                    (OpenBlock::Text, Delta::TextDelta { .. }) => {}
                    (OpenBlock::Tool(tool), Delta::InputJsonDelta { partial_json }) => {
                        tool.fragments.push_str(partial_json);
                        tool.fragment_count += 1;
                    }
                    _ => {
                        return Err(violate(
                            i,
                            format!(
                                "delta type {} does not match the open block kind",
                                delta_kind(delta)
                            ),
                        ));
                    }
                }
            }
            Event::ContentBlockStop { index } => {
                let Some((open_index, kind)) = open.take() else {
                    return Err(violate(i, "content_block_stop without an open block"));
                };
                if *index != open_index {
                    return Err(violate(
                        i,
                        format!(
                            "content_block_stop index {index} does not match the open block {open_index}"
                        ),
                    ));
                }
                if let OpenBlock::Tool(tool) = kind {
                    validate_tool_input(i, &tool, allow_synthesized)?;
                }
            }
            Event::MessageDelta {
                stop_reason,
                stop_sequence,
                ..
            } => {
                if phase != Phase::Blocks {
                    return Err(violate(
                        i,
                        "message_delta outside the block region (missing message_start or out of grammar order)",
                    ));
                }
                if open.is_some() {
                    return Err(violate(i, "message_delta with an open block"));
                }
                if stop_sequence.is_some() && *stop_reason != StopReason::StopSequence {
                    return Err(violate(
                        i,
                        "stop_sequence is only permitted when stop_reason is stop_sequence",
                    ));
                }
                phase = Phase::NeedMessageDone;
            }
            Event::MessageDone {} => {
                if phase == Phase::Done {
                    return Err(violate(i, "event after message_done"));
                }
                if phase != Phase::NeedMessageDone {
                    return Err(violate(
                        i,
                        "message_done without an immediately preceding message_delta",
                    ));
                }
                phase = Phase::Done;
            }
        }
    }

    if phase == Phase::Done {
        return Ok(());
    }
    if let Some((index, _)) = open {
        return Err(violate(
            stream.events.len(),
            format!("block index {index} is not stopped"),
        ));
    }
    if phase == Phase::NeedMessageDone {
        return Err(violate(
            stream.events.len(),
            format!(
                "events: missing message_done (stream ends after {} events)",
                stream.events.len()
            ),
        ));
    }
    Err(violate(
        stream.events.len(),
        "events: missing message_delta",
    ))
}

fn validate_tool_input(
    i: usize,
    tool: &OpenTool,
    allow_synthesized: bool,
) -> Result<(), Violation> {
    if tool.fragment_count == 0 {
        if tool.input.is_empty() || allow_synthesized {
            return Ok(());
        }
        return Err(violate(
            i,
            "tool block input without input_json_delta fragments; only encoder shorthand may synthesize them",
        ));
    }
    if tool.input != tool.fragments {
        return Err(violate(
            i,
            format!(
                "tool block input does not equal the concatenation of its {} fragments (INV-1 exact text)",
                tool.fragment_count
            ),
        ));
    }
    Ok(())
}

fn block_kind(block: &Block) -> &'static str {
    match block {
        Block::Text { .. } => "text",
        Block::Image { .. } => "image",
        Block::ToolUse { .. } => "tool_use",
        Block::ToolResult { .. } => "tool_result",
    }
}

fn delta_kind(delta: &Delta) -> &'static str {
    match delta {
        Delta::TextDelta { .. } => "text_delta",
        Delta::InputJsonDelta { .. } => "input_json_delta",
    }
}

File: python/src/oxa/ir/checker.py
"""Event-stream invariant checking (spec/01 §7).

Enforces INV-5 grammar, INV-6 index discipline, and the relational rule that a
tool block's input equals the exact concatenation of its
input_json_delta.partial_json fragments (INV-1: compared as opaque strings).
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum

from oxa.ir.constants import STOP_STOP_SEQUENCE
from oxa.ir.types import (
    ContentBlockDelta,
    ContentBlockStart,
    ContentBlockStop,
    EventStream,
    InputJsonDelta,
    MessageDelta,
    MessageDone,
    MessageStart,
    TextBlock,
    TextDelta,
    ToolUseBlock,
)


@dataclass(frozen=True, slots=True)
class Violation(Exception):
    """A rule violation found by the event-stream validator."""

    event: int
    message: str

    def __str__(self) -> str:
        return f"event {self.event}: {self.message}"


class _Phase(Enum):
    NEED_START = 1
    BLOCKS = 2
    NEED_MESSAGE_DONE = 3
    DONE = 4


@dataclass(slots=True)
class _OpenTool:
    input: str
    fragments: list[str]
    fragment_count: int


def validate_event_stream(stream: EventStream) -> None:
    """Validates a decoder-produced stream (strict).

    A tool block with non-empty input must carry explicit input fragments.
    """
    _validate_with(stream, allow_synthesized=False)


def validate_event_stream_for_encoder(stream: EventStream) -> None:
    """Validates an encoder-input stream (lenient).

    Accepts the documented encoder shorthand where a tool block carries its
    full input and no fragments (the encoder synthesizes one full delta).
    """
    _validate_with(stream, allow_synthesized=True)


def _validate_with(stream: EventStream, allow_synthesized: bool) -> None:
    phase = _Phase.NEED_START
    next_index = 0
    # Currently open block: (index, "text" | _OpenTool)
    open_block: tuple[int, str | _OpenTool] | None = None

    for i, event in enumerate(stream.events):
        if isinstance(event, MessageStart):
            if phase != _Phase.NEED_START:
                raise Violation(i, "duplicate message_start")
            phase = _Phase.BLOCKS

        elif isinstance(event, ContentBlockStart):
            if phase != _Phase.BLOCKS:
                raise Violation(
                    i,
                    "content_block_start outside the block region (no message_start, or a message_delta already seen)",
                )
            if open_block is not None:
                raise Violation(i, "content_block_start with an open block")
            if event.index != next_index:
                raise Violation(
                    i,
                    f"content_block_start index {event.index}, want {next_index}",
                )
            next_index += 1

            if isinstance(event.block, TextBlock):
                open_block = (event.index, "text")
            elif isinstance(event.block, ToolUseBlock):
                open_block = (
                    event.index,
                    _OpenTool(
                        input=event.block.input,
                        fragments=[],
                        fragment_count=0,
                    ),
                )
            else:
                block_kind = getattr(event.block, "type", type(event.block).__name__)
                raise Violation(
                    i,
                    f"content_block_start carries {block_kind} block; streams carry text and tool_use only",
                )

        elif isinstance(event, ContentBlockDelta):
            if open_block is None:
                raise Violation(i, "content_block_delta without an open block")
            open_idx, kind = open_block
            if event.index != open_idx:
                raise Violation(
                    i,
                    f"content_block_delta index {event.index} does not match the open block {open_idx}",
                )
            if kind == "text":
                if not isinstance(event.delta, TextDelta):
                    raise Violation(
                        i,
                        f"delta type {getattr(event.delta, 'type', type(event.delta).__name__)} does not match the open block kind",
                    )
            elif isinstance(kind, _OpenTool):
                if not isinstance(event.delta, InputJsonDelta):
                    raise Violation(
                        i,
                        f"delta type {getattr(event.delta, 'type', type(event.delta).__name__)} does not match the open block kind",
                    )
                kind.fragments.append(event.delta.partial_json)
                kind.fragment_count += 1

        elif isinstance(event, ContentBlockStop):
            if open_block is None:
                raise Violation(i, "content_block_stop without an open block")
            open_idx, kind = open_block
            open_block = None
            if event.index != open_idx:
                raise Violation(
                    i,
                    f"content_block_stop index {event.index} does not match the open block {open_idx}",
                )
            if isinstance(kind, _OpenTool):
                _validate_tool_input(i, kind, allow_synthesized)

        elif isinstance(event, MessageDelta):
            if phase != _Phase.BLOCKS:
                raise Violation(
                    i,
                    "message_delta outside the block region (missing message_start or out of grammar order)",
                )
            if open_block is not None:
                raise Violation(i, "message_delta with an open block")
            if event.stop_sequence is not None and event.stop_reason != STOP_STOP_SEQUENCE:
                raise Violation(
                    i,
                    "stop_sequence is only permitted when stop_reason is stop_sequence",
                )
            phase = _Phase.NEED_MESSAGE_DONE

        elif isinstance(event, MessageDone):
            if phase == _Phase.DONE:
                raise Violation(i, "event after message_done")
            if phase != _Phase.NEED_MESSAGE_DONE:
                raise Violation(
                    i,
                    "message_done without an immediately preceding message_delta",
                )
            phase = _Phase.DONE

    if phase == _Phase.DONE:
        return

    total = len(stream.events)
    if open_block is not None:
        raise Violation(total, f"block index {open_block[0]} is not stopped")
    if phase == _Phase.NEED_MESSAGE_DONE:
        raise Violation(total, f"events: missing message_done (stream ends after {total} events)")
    raise Violation(total, "events: missing message_delta")


def _validate_tool_input(i: int, tool: _OpenTool, allow_synthesized: bool) -> None:
    if tool.fragment_count == 0:
        if not tool.input or allow_synthesized:
            return
        raise Violation(
            i,
            "tool block input without input_json_delta fragments; only encoder shorthand may synthesize them",
        )
    concatenated = "".join(tool.fragments)
    if tool.input != concatenated:
        raise Violation(
            i,
            f"tool block input does not equal the concatenation of its {tool.fragment_count} fragments (INV-1 exact text)",
        )
